torch/autograd/graph.py

```diff
@@ -269,3 +271,125 @@ def __setstate__(self, state):
         handles.append(t.register_hook(get_inner_hook(i)))
 
     return Handle(tuple(handles))
+
+
+# NOTE [Allow mutation on tensors saved for backward]
+#
+# 1. Tensor gets saved for backward
+#    - remember the python object id and the version of the tensor
+#    - remember aliasing information (data_ptr of base + version)
+#    - save the original so we control its lifetime
+# 2. Any time a tensor gets in-placed
+#    - for each tensor aliased to it:
+#      - check using its object id and version to see if it has been saved
+#      - if it has been saved, clone it
+#      - delete the reference to the original
+# 3. during backward
+#    - if the clone exists, the tensor must've been modified in-place
+
+_cloned: weakref.WeakKeyDictionary = weakref.WeakKeyDictionary()
+_original: weakref.WeakKeyDictionary = weakref.WeakKeyDictionary()
+_tid_to_weakhandle: Dict[Tuple[int, int, int], weakref.ReferenceType] = dict()
+_sid_to_tid: Dict[Tuple[int, int], Set[Tuple[int, int, int]]] = dict()
+_ctx_id = 0
+_inside_ctx = False
+
+def _get_tid(t) -> Tuple[int, int, int]:
+    return (id(t), t.data_ptr(), t._version)
+
+def _get_sid(t) -> Tuple[int, int]:
+    return (t.data_ptr(), t._version)
+
+class _Handle():
+    pass
+
+class _swap_with_cloned(saved_tensors_hooks):
+    def __init__(self):
+        def pack_hook(t):
+            tid = _get_tid(t)
+            sid = _get_sid(t)
+            # Tensors saved for backward have an entry in _tid_to_weakhandle
+            handle: Optional[_Handle] = None
+
+            # Save aliasing information
+            if sid not in _sid_to_tid:
+                _sid_to_tid[sid] = set()
+            _sid_to_tid[sid].add(tid)
+
+            # NB: The same tensor (of the same version) can be saved multiple times
+            if tid not in _tid_to_weakhandle or _tid_to_weakhandle[tid]() is None:
+                handle = _Handle()
+                _tid_to_weakhandle[tid] = weakref.ref(handle)
+                _original[handle] = t
+            else:
+                # Store an additional handle
+                handle = _tid_to_weakhandle[tid]()
+            return _ctx_id, handle
+
+        def unpack_hook(tup):
+            ctx_id, handle = tup
+            assert ctx_id == _ctx_id, (
+                "Trying to backward outside of the 'allow_mutation_on_saved_tensors' context"
+                "in which the graph was originally recorded.")
+            if handle in _cloned:
+                res = _cloned[handle]
+            else:
+                res = _original[handle]
+            return res
+
+        super().__init__(pack_hook, unpack_hook)
+
+class _CloneArgBeforeMutateMode(TorchDispatchMode):
+    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        if kwargs is None:
+            kwargs = {}
+
+        # Bug? Maybe something to do with wrapped number + dispatcher, not sure why yet
```

Comment:
Yes, Tensor which wrap plain python scalar do get unpacked before calling here.
But that should never happen for inplace ops as `torch.add_(2, torch.rand(3))` doesn't exist. In which case exactly does that happen?

Reply:
It happens when I try to print the tensor in the saved tensor hook:
- when I run `a *= 10`, 10 is saved as a wrapped number
- call pack hook, which prints the tensor, calling `t.reshape(-1)`
- go into torch_dispatch for `aten::_reshape_alias`, which turns wrapped number back into an int
- call `aten::_reshape_alias` with int as first argument, producing the error

Reply:
This seems to be an issue with saved variable hook. Wrapped number is a cpp thing, but because saved variable hook allows us to insert python code at this level. I think our options are:
- wrap and unwrap wrapped number (just like what torch dispatch does)
- detect if the tensor is a wrapped number and simply do not trigger saved variable hook

Reply:
> wrap and unwrap wrapped number (just like what torch dispatch does)
 
I guess we can't really do that because the attribute on the Node is already a SavedVariable. And we don't really want to add an extra field on the SavedVariable to be able to save either a Tensor or a number?

Is there an option 3: clone wrapped numbers to make them regular Tensors before triggering the SavedVariable logic?

Reply:
I feel like not triggering the hooks for wrapped tensor is more expected. In the user's perspective that number was never a tensor, it's only a tensor in the backend for convenience.

Is there a reason allowing the user to still be able to interpose on wrapped numbers is preferred?

Reply:
Not triggering the hook sounds ok!